File: app/src/main/java/com/example/myapplication/view/auth/SignUpInputEmailFragment.java
package com.example.myapplication.view.auth;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.example.myapplication.R;
import com.example.myapplication.databinding.FragmentSignUpInputEmailBinding;
import com.example.myapplication.viewmodel.SignUpViewModel;
import com.mobsandgeeks.saripaar.ValidationError;
import com.mobsandgeeks.saripaar.Validator;
import com.mobsandgeeks.saripaar.annotation.Email;

import java.util.List;


public class SignUpInputEmailFragment extends Fragment implements Validator.ValidationListener {
    FragmentSignUpInputEmailBinding binding;
    SignUpViewModel signUpViewModel;

    @Email(message = "이메일 형식으로")
    EditText inputEmail;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        // 뷰 인플레이트
        binding = FragmentSignUpInputEmailBinding.inflate(inflater, container, false);

        inputEmail = binding.inputEmail;


        signUpViewModel = new ViewModelProvider(requireActivity()).get(SignUpViewModel.class);


        Validator validator = new Validator(this);
        validator.setValidationListener(this);


        binding.inputEmail.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                validator.validate();
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                validator.validate();
            }

            @Override
            public void afterTextChanged(Editable s) {
                validator.validate();
            }
        });

        binding.nextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                binding.prg.setProgress(0);

                signUpViewModel.setEmail(binding.inputEmail.getText().toString());// 사용자가 입력한 이메일을 뷰모델에 저장

                Navigation.findNavController(v).navigate(R.id.action_signUpInputEmailFragment2_to_signUpInputPasswordFragment);
            }
        });

        return binding.getRoot();

    }

    //유효성 검사 통과시 호출
    @Override
    public void onValidationSucceeded() {
        // 이메일 형식이 유효할 때 버튼 배경 변경 및 활성화
        binding.nextButton.setBackgroundResource(R.drawable.btn_black);
        binding.nextButton.setEnabled(true);
    }

    //유효성 검사 오류시 호출
    @Override
    public void onValidationFailed(List<ValidationError> errors) {
        // 이메일 형식이 유효하지 않을 때 버튼 배경 원래대로 변경 및 비활성화
        binding.nextButton.setBackgroundResource(R.drawable.btn_gray);
        binding.nextButton.setEnabled(false);
    }

}

File: app/src/main/java/com/example/myapplication/view/auth/SignUpInputEmailFragment.kt
package com.example.myapplication.view.auth

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.navigation.Navigation
import com.example.data.api.AuthApi
import com.example.myapplication.R
import com.example.myapplication.databinding.FragmentSignUpInputEmailBinding
import com.example.myapplication.viewmodel.AuthViewModel
import com.mobsandgeeks.saripaar.ValidationError
import com.mobsandgeeks.saripaar.Validator
import com.mobsandgeeks.saripaar.annotation.Email
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class SignUpInputEmailFragment : Fragment(), Validator.ValidationListener {

    @Inject
    lateinit var authApi: AuthApi

    private lateinit var binding: FragmentSignUpInputEmailBinding
    private lateinit var authViewModel: AuthViewModel

    @Email(message = "이메일 형식이 유효하지 않습니다.")
    lateinit var inputEmail: EditText

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentSignUpInputEmailBinding.inflate(inflater, container, false)
        authViewModel = ViewModelProvider(requireActivity())[AuthViewModel::class.java]
        binding.viewModel = authViewModel
        inputEmail = binding.inputEmail


        val validator = Validator(this)
        validator.setValidationListener(this)
        binding.inputEmail.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                validator.validate()
            }
        })

        binding.nextButton.setOnClickListener {
            lifecycleScope.launch {
                val response = authApi.emailDuplicateCheck(authViewModel.email.value!!)
                if (response.isSuccessful) {
                    if (response.body()!!.isAvailable) {
                        Toast.makeText(context, "사용 가능한 이메일입니다.", Toast.LENGTH_SHORT).show()
                        authViewModel.email.value = binding.inputEmail.text.toString().trim()
                        Navigation.findNavController(binding.root)
                            .navigate(R.id.action_signUpInputEmailFragment2_to_signUpInputPasswordFragment)
                    } else {
                        Toast.makeText(context, "이미 사용중인 이메일입니다.", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    Toast.makeText(context, "이메일 중복 확인에 실패했습니다.", Toast.LENGTH_SHORT).show()
                }
            }
        }

        return binding.root
    }

    override fun onValidationFailed(errors: MutableList<ValidationError>) {
        binding.nextButton.setBackgroundResource(R.drawable.btn_gray)
        binding.nextButton.isEnabled = false
    }

    override fun onValidationSucceeded() {
        binding.nextButton.setBackgroundResource(R.drawable.btn_black)
        binding.nextButton.isEnabled = true
    }
}
